src/eve/pattern_matching.py

```diff
@@ -0,0 +1,159 @@
+    elif a != b:
+        yield (path, f"Values are not equal. `{a}` != `{b}`")
+
+
+@get_differences.register
+def _(a: ObjectPattern, b: Any, path: str = "") -> Iterator[tuple[str, str]]:
+    if not isinstance(b, a.cls):
+        yield (
+            path,
+            f"Expected an instance of class {a.cls.__name__}, but got {type(b).__name__}",
+        )
+    else:
+        for k in a.attrs.keys():
+            if not hasattr(b, k):
+                yield (path, f"Value has no attribute {k}.")
+            else:
+                yield from get_differences(a.attrs[k], getattr(b, k), path=f"{path}.{k}")
+
+
+@get_differences.register
+def _(a: list, b: Any, path: str = "") -> Iterator[tuple[str, str]]:
+    if not isinstance(b, list):
+        yield (path, f"Expected list, but got {type(b).__name__}")
+    elif len(a) != len(b):
+        yield (path, f"Expected list of length {len(a)}, but got length {len(b)}")
+    else:
+        for i, (el_a, el_b) in enumerate(zip(a, b)):
+            for diff in get_differences(el_a, el_b, path=f"{path}[{i}]"):
+                yield diff
+
+
+@get_differences.register
+def _(a: dict, b: Any, path: str = "") -> Iterator[tuple[str, str]]:
+    if not isinstance(b, dict):
+        yield (path, f"Expected dict, but got {type(b).__name__}")
+    elif set(a.keys()) != set(b.keys()):
+        a_min_b = set(a.keys()).difference(b.keys())
+        b_min_a = set(b.keys()).difference(a.keys())
+        if a_min_b:
+            missing_keys_str = "`" + "`, `".join(map(str, a_min_b)) + "`"
+            yield (
+                path,
+                f"Expected dictionary with keys `{'`, `'.join(map(str, a.keys()))}`, but the following keys are missing: {missing_keys_str}",
+            )
+        if b_min_a:
+            extra_keys_str = "`" + "`, `".join(map(str, b_min_a)) + "`"
+            yield (
+                path,
+                f"Expected dictionary with keys `{'`, `'.join(map(str, a.keys()))}`, but the following keys are extra: {extra_keys_str}",
+            )
+    else:
+        for k, v_a, v_b in zip(a.keys(), a.values(), b.values()):
+            yield from get_differences(v_a, v_b, path=f'{path}["{k}"]')
+
+
+@dataclass(frozen=True)
+class ModuleWrapper:
+    """
+    Small wrapper to conveniently create `ObjectPattern`s for classes of a module.
+
+    Example:
+    >>> import foo_ir  # doctest: +SKIP
+    >>> foo_ir_ = ModuleWrapper(foo_ir)  # doctest: +SKIP
+    >>> assert foo_ir_.Foo(bar="baz").match(foo_ir.Foo(bar="baz", foo="bar"))  # doctest: +SKIP
+    >>> assert not foo_ir_.Foo(bar="bar").match(foo_ir.Foo(bar="baz", foo="bar"))  # doctest: +SKIP
+    """
+
+    module: ModuleType
+
+    def __getattr__(self, item: str) -> _ObjectPatternConstructor:
+        val = getattr(self.module, item)
+        if not inspect.isclass(val):
+            raise ValueError("Only classes allowed.")
+        return _ObjectPatternConstructor(val)
```

Comment:
As I said in my previous review, I think this util causes more harm than good since makes the creation of pattern objects look exactly the same as instantiation actual objects. I think a more explicit approach should be used here

Reply:
I missed 3 comments from your last review including this one. As discussed I see your point, but still like this solution better. I'll let this rest for a couple of days to see what I think and ask Rico and then we can revisit. Until then there is a comment and we should merge if the rest is sorted out.